File: backend/websocket/handler.py
from datetime import datetime, timezone
from fastapi import APIRouter, WebSocket, WebSocketDisconnect
from auth import decode_access_token
from database import get_db
from websocket.manager import manager
from hpo.message_queue import enqueue_message
import config

router = APIRouter()


@router.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket, token: str = ""):
    # Authenticate via JWT token query parameter
    payload = decode_access_token(token)
    if not payload:
        await websocket.close(code=4001, reason="Invalid token")
        return

    user_id = payload.get("sub")
    if not user_id:
        await websocket.close(code=4001, reason="Invalid token")
        return

    await websocket.accept()
    await manager.connect(websocket, user_id)
    await websocket.send_json({"type": "auth:success", "userId": user_id})

    print(f"[WS] User connected: {user_id}")

    try:
        while True:
            data = await websocket.receive_json()
            msg_type = data.get("type")

            if msg_type == "conversation:join":
                conv_id = data.get("conversationId") or data.get("conversation_id")
                if conv_id:
                    manager.join_room(user_id, conv_id)

            elif msg_type == "message:send":
                await handle_message_send(user_id, data)

            elif msg_type == "user:typing":
                conv_id = data.get("conversationId") or data.get("conversation_id")
                if conv_id:
                    await manager.broadcast_to_room(conv_id, {
                        "type": "user:typing",
                        "userId": user_id,
                        "conversationId": conv_id,
                    }, exclude=user_id)

            elif msg_type == "user:stopTyping":
                conv_id = data.get("conversationId") or data.get("conversation_id")
                if conv_id:
                    await manager.broadcast_to_room(conv_id, {
                        "type": "user:stopTyping",
                        "userId": user_id,
                        "conversationId": conv_id,
                    }, exclude=user_id)

    except WebSocketDisconnect:
        pass
    except Exception as e:
        print(f"[WS] Error for user {user_id}: {e}")
    finally:
        await manager.disconnect(user_id)
        print(f"[WS] User disconnected: {user_id}")


async def handle_message_send(sender_id: str, data: dict):
    db = get_db()
    msg_data = data.get("data", {})
    request_id = data.get("requestId") or data.get("request_id")

    try:
        now = datetime.now(timezone.utc)

        # Store message in MongoDB
        msg_doc = {
            "conversation_id": msg_data.get("conversationId"),
            "sender_id": sender_id,
            "recipient_id": msg_data.get("recipientId"),
            "ciphertext": msg_data.get("ciphertext", ""),
            "ratchet_header": msg_data.get("ratchetHeader", ""),
            "timestamp": now,
            "delivered": False,
        }
        result = await db.messages.insert_one(msg_doc)
        message_id = str(result.inserted_id)

        # Update conversation timestamp
        from bson import ObjectId
        try:
            await db.conversations.update_one(
                {"_id": ObjectId(msg_data.get("conversationId"))},
                {"$set": {"updated_at": now}},
            )
        except Exception:
            pass

        # Build the packet to send to recipient
        packet = {
            **msg_data,
            "id": message_id,
            "timestamp": int(now.timestamp() * 1000),
        }

        recipient_id = msg_data.get("recipientId")

        if config.HPO_ENABLED and manager.is_online(recipient_id):
            # Route through HPO message queue
            enqueue_message(recipient_id, {
                "type": "real",
                "packet": packet,
                "recipientId": recipient_id,
            })
        elif manager.is_online(recipient_id):
            # Direct delivery
            await manager.send_to_user(recipient_id, {
                "type": "message:receive",
                "data": packet,
            })

        # Acknowledge to sender
        await manager.send_to_user(sender_id, {
            "type": "message:ack",
            "requestId": request_id,
            "success": True,
            "messageId": message_id,
        })

    except Exception as e:
        print(f"[WS] Error saving message: {e}")
        await manager.send_to_user(sender_id, {
            "type": "message:ack",
            "requestId": request_id,
            "success": False,
        })

File: src/lib/ws/handler.ts
import type { WebSocket } from "ws";
import { ObjectId } from "mongodb";
import { getDb } from "@/lib/db";
import { decodeAccessToken } from "@/lib/auth-server";
import { manager } from "./manager";
import { isHpoEnabled } from "@/lib/hpo/config";
import { enqueueMessage } from "@/lib/hpo/message-queue";

export async function handleConnection(ws: WebSocket, token: string): Promise<void> {
  const payload = await decodeAccessToken(token);
  if (!payload?.sub) {
    ws.close(4001, "Invalid token");
    return;
  }

  const userId = payload.sub;
  await manager.connect(ws, userId);
  ws.send(JSON.stringify({ type: "auth:success", userId }));
  console.log(`[WS] User connected: ${userId}`);

  ws.on("message", async (raw) => {
    let data: Record<string, unknown>;
    try {
      data = JSON.parse(raw.toString());
    } catch {
      return;
    }

    const msgType = data.type as string;

    if (msgType === "conversation:join") {
      const convId = (data.conversationId ?? data.conversation_id) as
        | string
        | undefined;
      if (convId) manager.joinRoom(userId, convId);
    } else if (msgType === "message:send") {
      await handleMessageSend(userId, data);
    } else if (msgType === "user:typing" || msgType === "user:stopTyping") {
      const convId = (data.conversationId ?? data.conversation_id) as
        | string
        | undefined;
      if (convId) {
        manager.broadcastToRoom(
          convId,
          { type: msgType, userId, conversationId: convId },
          userId
        );
      }
    }
  });

  ws.on("close", () => {
    manager.disconnect(userId);
    console.log(`[WS] User disconnected: ${userId}`);
  });

  ws.on("error", (err) => {
    console.error(`[WS] Error for user ${userId}:`, err);
  });
}

async function handleMessageSend(
  senderId: string,
  data: Record<string, unknown>
): Promise<void> {
  const msgData = (data.data ?? {}) as Record<string, unknown>;
  const requestId = (data.requestId ?? data.request_id) as string | undefined;

  try {
    const db = await getDb();
    const now = new Date();
    const conversationId = msgData.conversationId as string | undefined;
    const recipientId = msgData.recipientId as string | undefined;

    const result = await db.collection("messages").insertOne({
      conversation_id: conversationId,
      sender_id: senderId,
      recipient_id: recipientId,
      ciphertext: (msgData.ciphertext as string) ?? "",
      ratchet_header: (msgData.ratchetHeader as string) ?? "",
      timestamp: now,
      delivered: false,
    });
    const messageId = result.insertedId.toString();

    if (conversationId) {
      try {
        await db
          .collection("conversations")
          .updateOne(
            { _id: new ObjectId(conversationId) },
            { $set: { updated_at: now } }
          );
      } catch {
        // Invalid ObjectId — ignore
      }
    }

    const packet = {
      ...msgData,
      id: messageId,
      timestamp: now.getTime(),
    };

    if (recipientId && isHpoEnabled() && manager.isOnline(recipientId)) {
      enqueueMessage(recipientId, { type: "real", packet, recipientId });
    } else if (recipientId && manager.isOnline(recipientId)) {
      manager.sendToUser(recipientId, { type: "message:receive", data: packet });
    }

    manager.sendToUser(senderId, {
      type: "message:ack",
      requestId,
      success: true,
      messageId,
    });
  } catch (err) {
    console.error("[WS] Error saving message:", err);
    manager.sendToUser(senderId, {
      type: "message:ack",
      requestId,
      success: false,
    });
  }
}
